presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/JtsGeometryUtils.java

```diff
@@ -0,0 +1,474 @@
+        skipEnvelope(input);
+        int partCount = input.readInt();
+        if (partCount == 0) {
+            return GEOMETRY_FACTORY.createPolygon();
+        }
+
+        int pointCount = input.readInt();
+        int[] startIndexes = new int[partCount];
+        for (int i = 0; i < partCount; i++) {
+            startIndexes[i] = input.readInt();
+        }
+
+        int[] partLengths = new int[partCount];
+        if (partCount > 1) {
+            partLengths[0] = startIndexes[1];
+            for (int i = 1; i < partCount - 1; i++) {
+                partLengths[i] = startIndexes[i + 1] - startIndexes[i];
+            }
+        }
+        partLengths[partCount - 1] = pointCount - startIndexes[partCount - 1];
+
+        LinearRing shell = null;
+        List<LinearRing> holes = new ArrayList<>();
+        List<Polygon> polygons = new ArrayList<>();
+        for (int i = 0; i < partCount; i++) {
+            Coordinate[] coordinates = readCoordinates(input, partLengths[i]);
+            if (!isClockwise(coordinates)) {
+                holes.add(GEOMETRY_FACTORY.createLinearRing(coordinates));
+                continue;
+            }
+            if (shell != null) {
+                polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
+            }
+            shell = GEOMETRY_FACTORY.createLinearRing(coordinates);
+            holes.clear();
+        }
+        polygons.add(GEOMETRY_FACTORY.createPolygon(shell, holes.toArray(new LinearRing[0])));
+
+        if (polygons.size() == 1) {
+            return polygons.get(0);
+        }
+        return GEOMETRY_FACTORY.createMultiPolygon(polygons.toArray(new Polygon[0]));
+    }
+
+    private static void writePolygon(Geometry geometry, SliceOutput output) {
+        requireNonNull(geometry, "polygon is null");
+        requireNonNull(output, "output is null");
+        if (!geometry.getGeometryType().equals(JTS_POLYGON) && !geometry.getGeometryType().equals(JTS_MULTI_POLYGON)) {
+            throw new IllegalArgumentException("writePolyline can only handle multi-polygon or polygon");
+        }
+
+        int numGeometries = geometry.getNumGeometries();
+        int numParts = 0;
+        int numPoints = geometry.getNumPoints();
+        for (int i = 0; i < numGeometries; i++) {
+            Polygon polygon = (Polygon) geometry.getGeometryN(i);
+            if (polygon.getNumPoints() > 0) {
+                numParts += polygon.getNumInteriorRing() + 1;
+            }
+        }
+
+        int size = SIZE_OF_INT + SIZE_OF_DOUBLE * 4 + SIZE_OF_INT * (2 + numParts)
+                + SIZE_OF_DOUBLE * 2 * numPoints;
+        output.writeInt(size);
+
+        output.writeInt(EsriShapeType.POLYGON.code);
+
+        writeEnvelope(geometry.getEnvelopeInternal(), output);
+
+        output.writeInt(numParts);
+
+        output.writeInt(numPoints);
+
+        // Index of first point in part
+        int startIndex[] = new int[numParts + 1];
```

Comment:
drop this array like you did in writePolyline and write out the lengths as you go